File: caminho-minimo/c++/floyd-warshall.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

// Função para imprimir a matriz de distâncias
void printMatrix(const vector<vector<int>>& matrix) {
    int numVertices = matrix.size();
    cout << "Matriz de distâncias:" << endl;
    for (int i = 0; i < numVertices; ++i) {
        for (int j = 0; j < numVertices; ++j) {
            if (matrix[i][j] == INT_MAX)
                cout << "INF\t";
            else
                cout << matrix[i][j] << "\t";
        }
        cout << endl;
    }
}

// Função que implementa o algoritmo Floyd-Warshall
void floydWarshall(vector<vector<int>>& graph) {
    int numVertices = graph.size();

    // Inicializa a matriz de distâncias com os pesos das arestas
    vector<vector<int>> distances(graph);

    // Atualiza as distâncias considerando todos os nós intermediários
    for (int k = 0; k < numVertices; ++k) {
        for (int i = 0; i < numVertices; ++i) {
            for (int j = 0; j < numVertices; ++j) {
                // Verifica se a rota passando por k é mais curta
                if (distances[i][k] != INT_MAX && distances[k][j] != INT_MAX &&
                    distances[i][k] + distances[k][j] < distances[i][j]) {
                    distances[i][j] = distances[i][k] + distances[k][j];
                }
            }
        }
    }

    printMatrix(distances);
}

// Exemplo de uso
int main() {
    // Cria um grafo de exemplo (matriz de adjacência)
    vector<vector<int>> graph = {
        {   0   ,   5   ,INT_MAX,   10  },
        {INT_MAX,   0   ,   3   ,INT_MAX},
        {INT_MAX,INT_MAX,   0   ,   1   },
        {INT_MAX,INT_MAX,INT_MAX,   0   }  
    };

    floydWarshall(graph);

    return 0;
}

File: caminho-minimo/js/floyd-warshall.js
function floydWarshall(graph) {
  const numNodes = Object.keys(graph).length;
  const distances = {};
  const next = {};

  // Inicializa as matrizes de distâncias e próximo
  for (const source in graph) {
    distances[source] = {};
    next[source] = {};
    for (const destination in graph) {
      if (source === destination) {
        distances[source][destination] = 0;
      } else if (graph[source][destination] !== undefined) {
        distances[source][destination] = graph[source][destination];
      } else {
        distances[source][destination] = Infinity;
      }
      next[source][destination] = destination;
    }
  }

  // Algoritmo de Floyd-Warshall
  for (const k in graph) {
    for (const i in graph) {
      for (const j in graph) {
        const newDistance = distances[i][k] + distances[k][j];
        if (newDistance < distances[i][j]) {
          distances[i][j] = newDistance;
          next[i][j] = next[i][k];
        }
      }
    }
  }

  return { distances, next };
}

// Exemplo de uso:

const graph = {
  A: { B: 5, C: 2 },
  B: { A: 5, C: 1, D: 3 },
  C: { A: 2, B: 1, D: 1 },
  D: { B: 3, C: 1, E: 2 },
  E: { D: 2 },
};

const { distances, next } = floydWarshall(graph);

console.log("Distâncias entre todos os pares de nós:");
console.log(distances);

console.log("Próximo nó no caminho mais curto:");
console.log(next);
